packages/buidler-etherscan/src/index.ts

```diff
@@ -1,93 +1,196 @@
-
-task("verify-contract", "Verifies contract on etherscan")
-  .addParam("contractName", "Name of the deployed contract")
-  .addParam("address", "Deployed address of smart contract")
-  .addOptionalParam(
-    "libraries",
-    'Stringified JSON object in format of {library1: "0x2956356cd2a2bf3202f771f50d3d14a367b48071"}'
-  )
+import { ActionType, Artifact } from "@nomiclabs/buidler/types";
+// import {
+//   getVerificationStatus,
+//   verifyContract,
+// } from "./etherscan/EtherscanService";
+
+
+// To be removed?
+// import AbiEncoder from "./AbiEncoder";
+// import { toRequest } from "./etherscan/EtherscanVerifyContractRequest";
+
+
+const pluginName = "buidler-etherscan";
+
+
+interface VerificationArgs {
+  address: string;
+  constructorArguments: string[];
+  // Filename of constructor arguments module.
+  constructorArgs: string;
+};
+
+const verify: ActionType<VerificationArgs> = async (
+  { address, constructorArguments, constructorArgs: constructorArgsModule },
+  { config, network, run }
+) => {
+  const { getDefaultEtherscanConfig } = await import("./config");
+  const etherscan = getDefaultEtherscanConfig(config);
+
+  const { BuidlerPluginError } = await import("@nomiclabs/buidler/plugins");
+  if (etherscan.apiKey === undefined || etherscan.apiKey.trim() === "") {
+    // TODO: add URL to etherscan documentation?
+    throw new BuidlerPluginError(
+      pluginName,
+      "Please provide an Etherscan API token via buidler config. " +
+      "E.g.: { [...], etherscan: { apiKey: 'an API key' }, [...] }"
+    );
+  }
+
+  const { isAddress } = await import("@ethersproject/address");
+  if (!isAddress(address)) {
+    throw new BuidlerPluginError(pluginName, `${address} is an invalid address.`);
+  }
+
+  const { getVersionNumber, inferSolcVersion, InferralType } = await import("./solc/SolcVersions");
+  const solcVersionConfig = getVersionNumber(config.solc.version);
+
+  // Etherscan only supports solidity versions higher than or equal to v0.4.11.
+  // See https://etherscan.io/solcversions
+  // TODO: perhaps querying and scraping this list would be a better approach?
+  // This list should be validated - it links to https://github.com/ethereum/solc-bin/blob/gh-pages/bin/list.txt
+  // which has many old compilers included in the list too.
+  if (
+    (solcVersionConfig.major == 0 && solcVersionConfig.minor == 4 && solcVersionConfig.patch < 11) ||
+    (solcVersionConfig.major == 0 && solcVersionConfig.minor < 4)
+  ) {
+    throw new BuidlerPluginError(
+      pluginName,
+      "Etherscan only supports compiler versions 0.4.11 and higher.\n" +
+      "See https://etherscan.io/solcversions for more information."
+    );
+  }
+
+  //TODO: import constructorArgsModule if it is a valid file.
+
+  let etherscanAPIEndpoint: URL;
+  let deployedContractBytecode;
+  const {
+    getEtherscanEndpoint,
+    retrieveContractBytecode,
+    NetworkProberError,
+  } = await import("./network/prober");
+  try {
```

Comment:
I was actually trying to avoid throwing `BuidlerPluginError` inside the modules since they don't have enough context to show useful information for the user sometimes. The second option would work however. In general, I don't like the `null` as a return value but I do think that having too many try/catch constructs is a bit ugly. I'll see what I can do to minimize them.

Reply:
Update: `BuidlerPluginError`s are used for expected errors. In any case, I removed some of the custom errors to reduce the try/catch syntax in the task.